File: prset2/readability/readability.c
#include <ctype.h>
#include <cs50.h>
#include <stdio.h>
#include <string.h>
#include <math.h>

void printGrade(float index);
float clIndex(string text);
float lettersPer100Words(string text);
float sentencesPer100Words(string text);
int numLetters(string text);
int numWords(string text);
int numSentences(string text);

int main(void)
{
    string text = get_string("Text: ");
    printGrade(clIndex(text));
}

// printGrade : Float -> _
// Determines the grade level associated with a particular Coleman-Liau index
void printGrade(float index)
{
    int gradeLevel = round(index);

    if (gradeLevel >= 16)
    {
        printf("Grade 16+\n");
    }
    else if (gradeLevel < 1)
    {
        printf("Before Grade 1\n");
    }
    else
    {
        printf("Grade %i\n", gradeLevel);
    }
}

float clIndex(string text)
{
    return 0.0588 * lettersPer100Words(text) - 0.296 * sentencesPer100Words(text) - 15.8;
}

// lettersPer100Words : String -> Float
// Computes the average number of letters per 100 words of _text_
float lettersPer100Words(string text)
{
    return ((float) numLetters(text) / (float) numWords(text)) * 100;
}

// sentencesPer100Words : String -> Float
// Comutes the avarage number of sentences per 100 words of _text_
float sentencesPer100Words(string text)
{
    return ((float) numSentences(text) / (float) numWords(text)) * 100;
}

// numLetters : String -> Int
// Computes the number of letters (a-z or A-Z) in _text_
int numLetters(string text)
{
    int total = 0;

    // Iterate over the characters in the string
    for (int i = 0, length = strlen(text); i < length; i++)
    {
        // If the character is a letter, add it to the total
        if (isupper(text[i]) || islower(text[i]))
        {
            total++;
        }
        // Else, continue to the next character
    }

    return total;
}

// numWords : String -> Int
// Computes the number of words (any sequence of characters separated by spaces) in _text_
int numWords(string text)
{
    int total = 0;

    // Idea: Count the # of spaces, plus 1? (using a loop, increment whenever the character is a space)
    for (int i = 0, length = strlen(text); i < length; i++)
    {
        // Count all words after the first (' ' indicates start of new word)
        if (text[i] == ' ')
        {
            total++;
        }
        // Count the first word
        else if (i == 0)
        {
            total++;
        }
    }

    return total;
}


// numSentences : String -> Int
// Computes the number of sentences (any sequence of characters ending with a period, exclamation point, or question mark) in _text
int numSentences(string text)
{
    int total = 0;

    // Idea: Count the # of ending punctuation (using a loop)
    for (int i = 0, length = strlen(text); i < length; i++)
    {
        if (text[i] == '.' || text[i] == '!' || text[i] == '?')
        {
            total++;
        }
    }

    return total;

}

File: pset6/readability/readability.py
import re
from cs50 import get_string


def main():
    text = get_string("Text: ")
    print_grade(cl_index(text))


# print_grade : Float -> _
# Determines the grade level associated with a particular Coleman-Liau _index_


def print_grade(index):

    grade_level = round(index)

    if (grade_level >= 16):
        print("Grade 16+")
    elif (grade_level < 1):
        print("Before Grade 1")
    else:
        print(f"Grade {grade_level}")


# cl_index : String -> Float
# Determines the cl index for some _text_


def cl_index(text):
    return 0.0588 * letter_avg(text) - 0.296 * sentence_avg(text) - 15.8


# letter_avg : String -> Float
# Computes the avareage number of letters per every 100 words in _text_


def letter_avg(text):
    return 100 * num_letters(text) / num_words(text)


# sentence_avg : String -> Float
# Computers the average number of sentences per every 100 words in _text_


def sentence_avg(text):
    return 100 * num_sentences(text) / num_words(text)


# num_letters : String -> NN
# Computes the total number of letters in _text_


def num_letters(text):
    letters = re.findall("[A-Za-z]", text)
    return len(letters)


# num_words : String -> NN
# Computes the total number of words in _text_


def num_words(text):
    spaces = re.findall(" ", text)

    # Count the first word
    if (len(text)):
        return len(spaces) + 1

    return 0


# num_sentences : String -> NN
# Computes the total number of sentences in _text_


def num_sentences(text):
    sentences = re.findall("[.!?]", text)
    return len(sentences)


main()